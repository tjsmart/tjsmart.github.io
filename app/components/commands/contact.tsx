import { Command, registerCommand } from "./factory";

type Contact = {
    kind: string;
    username: string;
    href: string;
    icon: string;
};

const contacts: Contact[] = [
    {
        kind: "email",
        username: "tjsmart@ucsc.edu",
        href: "mailto:tjsmart@ucsc.edu",
        icon: "envelope",
    },
    {
        kind: "github",
        username: "tjsmart",
        href: "https://www.github.com/tjsmart",
        icon: "github",
    },
    {
        kind: "linkedin",
        username: "Tyler Smart",
        href: "https://www.linkedin.com/in/tyler-smart-0a4a419b",
        icon: "linkedin-square",
    },
    {
        kind: "google scholar",
        username: "Tyler Smart",
        href: "https://scholar.google.com/citations?user=r9supmkAAAAJ",
        icon: "graduation-cap",
    },
];

const contact = new Command(
    "contact",
    (args) => {
        if (args.length == 0) {
            return (
                <>
                    {contacts.map((contact, index) => (
                        <div key={index}>
                            <a
                                href={contact.href}
                                target="_blank"
                                rel="noopener noreferrer"
                                className="hover:text-red-500 underline"
                            >
                                <i className={`fa fa-${contact.icon}`}></i>
                                {` ${contact.kind} - ${contact.username}`}
                            </a>

                            <br />
                        </div>
                    ))}
                </>
            );
        } else if (args.length == 1) {
            const [arg] = args;
            let selectedContact = contacts.find(({ kind }) => kind == arg);
            if (selectedContact) {
                window.open(selectedContact.href, "_blank");
                return "";
            }
            return `error: unknown contact '${arg}'`;
        } else {
            return "error: contact takes at most a single argument";
        }
    },
    `How to contact/follow me, e.g., ${contacts
        .slice(0, 2)
        .map((contact) => contact.kind)
        .join(", ")}.`,
);

registerCommand(contact);

if (process.env["NODE_DEV"] == "TEST") {
    module.exports.contact = contact;
    module.exports.contacts = contacts;
}
