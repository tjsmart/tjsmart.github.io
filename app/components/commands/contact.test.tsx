import renderer from "react-test-renderer";
import { Command } from ".";
import * as mut from "./contact";

type Contact = {
    kind: string;
    username: string;
    href: string;
    icon: string;
};

// @ts-ignore
const contact: Command = mut.contact;
// @ts-ignore
const contacts: Contact[] = mut.contacts;

describe("contact", () => {
    test("contact snapshot", () => {
        const result = contact.callback([]);
        if (typeof result == "string") {
            fail("Expected result to not to be of type string");
        }

        const component = renderer.create(result);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("specific contacts opens href", () => {
        const mock_open = jest.fn();
        window.open = mock_open;

        contacts.forEach((c, index) => {
            contact.callback([c.kind]);
            expect(mock_open.mock.calls).toHaveLength(index + 1);
            expect(mock_open.mock.calls[index][0]).toStrictEqual(c.href);
        });
    });
});
