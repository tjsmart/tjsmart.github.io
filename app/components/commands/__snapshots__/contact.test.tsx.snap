// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact contact snapshot 1`] = `
[
  <div>
    <a
      className="hover:text-red-500 underline"
      href="mailto:tjsmart@ucsc.edu"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fa fa-envelope"
      />
       email - tjsmart@ucsc.edu
    </a>
    <br />
  </div>,
  <div>
    <a
      className="hover:text-red-500 underline"
      href="https://www.github.com/tjsmart"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fa fa-github"
      />
       github - tjsmart
    </a>
    <br />
  </div>,
  <div>
    <a
      className="hover:text-red-500 underline"
      href="https://www.linkedin.com/in/tyler-smart-0a4a419b"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fa fa-linkedin-square"
      />
       linkedin - Tyler Smart
    </a>
    <br />
  </div>,
  <div>
    <a
      className="hover:text-red-500 underline"
      href="https://scholar.google.com/citations?user=r9supmkAAAAJ"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fa fa-graduation-cap"
      />
       google scholar - Tyler Smart
    </a>
    <br />
  </div>,
]
`;
